### PR TITLE
Cache identifiers for better performance on getIriFromItem

### PR DESCRIPTION
**Pull request by soyuka**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | yes
| Deprecations? | no
| Tests pass?   | Not yet (still experimenting)
| Related tickets | #212
| License       | MIT
| Doc PR        | ø

## Add `getIriFromItem` cache

The collection here has ~8k items, no relations. Dev mode active but caching with apcu is forced.

[Blackfire before](https://blackfire.io/profiles/325404d0-c782-4f05-88ed-4676b6c3368b/graph):
- 2.87s spent in `getIriFromItem`
- 2.35 s for `getIdentifiersFromItem`

[Blackfire after](https://blackfire.io/profiles/7215dbe1-efb1-4835-aa8f-5296761c280a/graph):
- 898 ms spent in `getIriFromItem`
- 371 ms for `getIdentifiersFromItem`

When discussing my performance issue with @Simperfit we noticed that `getIdentifiersFromItem` was called a lot (previous method is `getIriFromItem`). This is called for every item in the collection.

### Profiler:

Don't look at the overall time, only the time spent serializing :p.

#### before
serializer3

#### after
cacheiri

In this PR, we cache the **properties** that are **identifiers** for the given class.

### CODE REVIEW

**Comment by teohhanhui** [comment]
I like the idea, but the implementation is wrong. We should decorate the IriConverter.

**Comment by soyuka** [comment]
👍 indeed

File: src/Bridge/Symfony/Routing/IriConverter.php [hunk 1]
```diff
@@ -112,6 +118,34 @@ private function getIdentifiersFromItem($item): array
         $identifiers = [];
         $resourceClass = $this->getObjectClass($item);
 
+        $cacheKey = self::CACHE_KEY_PREFIX.md5(serialize([$resourceClass]));
```
Comment (dunglas) [inline src/Bridge/Symfony/Routing/IriConverter.php]: Why not using `$resourceClass` directly as it's a string?

Comment (soyuka) [inline src/Bridge/Symfony/Routing/IriConverter.php]: Indeed, copy paste you know :dancing_men: 

I made this in a hurry just to see if my first assumption was correct! I'll refactor this and decorate the `IriConverter` as suggested by @teohhanhui. Will be done this week!

**Comment by soyuka** [comment]
Okay so I did this with a decorator, but because what's slow is `getIdentifiersFromItem` (which is a private method) I'm not sure it's ideal. Indeed, I had to copy the code from `IriConverter` (trying to avoid BC break at the same time), therefore I see no point in decorating :/.

You can check the code to understand what I mean... 

A solution would be to have a new service for `getIdentifiersFromItem` that we can cache without changing the IriConverter. Lmk if you've a better idea.

----- 

From my benchmarks, I know that we can still win a lot by caching the identifiers getters instead of calling the `PropertyAccessor` (`$identifiers[$propertyName] = $item->{$accessor}();`.

Some links (note that I'm using the normalizer dumper, 8000 items collection):

- [IriConverter without cache](https://blackfire.io/profiles/beab0865-7e50-40ed-823c-bc743e69edb3/graph) 8.36s
- [IriConverter with this PR cache](https://blackfire.io/profiles/51b859bd-956e-4f35-b5c5-79d3968717b7/graph) 6.89s
- [IriConverter with direct getter instead of PropertyAccessor + cache](https://blackfire.io/profiles/e9fda268-d964-4692-85cf-c85ed32391c9/graph) 4.78s

**Comment by teohhanhui** [comment]
@soyuka Okay, since what we're caching is a private method, then decoration indeed makes no sense.

> A solution would be to have a new service for getIdentifiersFromItem that we can cache without changing the IriConverter.

`ItemIdentifiersResolverInterface`?

**Comment by soyuka** [comment]
`ItemIdentifiersResolverInterface`? Extract is a bit strong no?

Would `get${ucfirst($propertyName)}` be acceptable to retrieve identifiers instead of using the property accessor?

**Comment by teohhanhui** [comment]
I think "extract" describes better what we're doing here: extracting identifiers from an item.

> Would get${ucfirst($propertyName)} be acceptable to retrieve identifiers instead of using the property accessor?

No, we cannot make such assumptions. Plus, it'd be a BC break.

**Comment by dunglas** [comment]
What we can do however is provide implementation of the `PropertyAccessorInterface` doing that. I usually do this for perf in my project. Such class should be included directly in Symfony.

**Comment by soyuka** [comment]
@dunglas You mean having a kind of IdentifierPropertyAccessor witch would end up doing something simpler than the original class (still no assumption could be made regarding `get{$attribute}` ? We also don't need `setValue` in this context.

**Comment by dunglas** [comment]
@soyuka yes, that you can register for your project if you stick to a particular convention (like public properties, or getters and setters). It prevents to trigger the introspection system, so it improves performances.

**Comment by soyuka** [comment]
Oh I'm doing this already on my project for this particular class :). So, I'll just add a `ItemIdentifiersExtractorInterface` so that identifiers can be cached here but I won't touch the propertyAccessor thing.

**Comment by soyuka** [comment]
ping @teohhanhui @dunglas 

I tried to avoid breaking changes lmk what you think. 
Note that Behat tests are green, just waiting for implementation approval to add tests (and #1014 comes first obviously).

File: src/Bridge/Symfony/Routing/ItemIdentifiersExtractor.php [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+namespace ApiPlatform\Core\Bridge\Symfony\Routing;
+
+use ApiPlatform\Core\Exception\RuntimeException;
+use ApiPlatform\Core\Metadata\Property\Factory\PropertyMetadataFactoryInterface;
+use ApiPlatform\Core\Metadata\Property\Factory\PropertyNameCollectionFactoryInterface;
+use ApiPlatform\Core\Util\ClassInfoTrait;
+use Symfony\Component\PropertyAccess\PropertyAccess;
+use Symfony\Component\PropertyAccess\PropertyAccessorInterface;
+
+class ItemIdentifiersExtractor implements ItemIdentifiersExtractorInterface
```
Comment (teohhanhui) [inline src/Bridge/Symfony/Routing/ItemIdentifiersExtractor.php]: Can be moved under `ApiPlatform\Core\Metadata\Property` namespace then?

File: src/Bridge/Symfony/Routing/CachedItemIdentifiersExtractor.php [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+{
+    use ClassInfoTrait;
+
+    const CACHE_KEY_PREFIX = 'iri_converter';
+
+    private $cacheItemPool;
+    private $propertyAccessor;
+    private $decorated;
+
+    public function __construct(CacheItemPoolInterface $cacheItemPool, PropertyAccessorInterface $propertyAccessor = null, ItemIdentifiersExtractorInterface $decorated = null)
```
Comment (Simperfit) [inline src/Bridge/Symfony/Routing/CachedItemIdentifiersExtractor.php]: should $decorated be null ?

File: src/Bridge/Symfony/Routing/ItemIdentifiersExtractor.php (hunk 2 above)
Comment (Simperfit) [inline src/Bridge/Symfony/Routing/ItemIdentifiersExtractor.php]: I agree

**Comment by Simperfit** [comment]
👍  Great work @soyuka

File: src/Bridge/Symfony/Routing/ItemIdentifiersExtractor.php (hunk 2 above)
Comment (dunglas) [inline src/Bridge/Symfony/Routing/ItemIdentifiersExtractor.php]: Indeed

File: src/Api/IdentifiersExtractor.php [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+            $relatedItem = $identifiers[$propertyName];
+
+            unset($identifiers[$propertyName]);
+
+            foreach ($this->propertyNameCollectionFactory->create($relatedResourceClass) as $relatedPropertyName) {
+                $propertyMetadata = $this->propertyMetadataFactory->create($relatedResourceClass, $relatedPropertyName);
+
+                if ($propertyMetadata->isIdentifier()) {
+                    if (isset($identifiers[$propertyName])) {
+                        throw new RuntimeException(sprintf('Composite identifiers not supported in "%s" through relation "%s" of "%s" used as identifier', $relatedResourceClass, $propertyName, $resourceClass));
```
Comment (soyuka) [inline src/Api/IdentifiersExtractor.php]: @teohhanhui I'm having a hard time trying to add test coverage on this one. IMO this is not a valid case, but maybe you can give me some more context/informations about the Exception? [Origin commit you authored](https://github.com/api-platform/core/commit/6117fe9786819f6b35ce0c9371bb30ab20a2637f).

**Comment by soyuka** [comment]
I did a great job covering this part as it wasn't tested anywhere (except through IriConverter, but not fully covered). 

There's still a point to discuss: the location of the `IdentifiersExtractor`. 

I'm not sure that it should actually be under `ApiPlatform\Core\Metadata` because it doesn't interfere with metadata, rather uses the metadata to extract identifiers. Therefore, IMHO this class should be in `ApiPlatform\Core\Api` (so is `ResourceClassResolver` for example). Let me know if you disagree we can always move it elsewhere :).

**Comment by soyuka** [comment]
All green ping @api-platform/core-team

**Comment by Simperfit** [comment]
@soyuka could you please update with master ?

File: src/Api/CachedIdentifiersExtractor.php [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    const CACHE_KEY_PREFIX = 'iri_identifiers';
+
+    private $cacheItemPool;
+    private $propertyAccessor;
+    private $decorated;
+
+    public function __construct(CacheItemPoolInterface $cacheItemPool, PropertyAccessorInterface $propertyAccessor = null, IdentifiersExtractorInterface $decorated)
+    {
+        $this->cacheItemPool = $cacheItemPool;
+        $this->propertyAccessor = $propertyAccessor ?: PropertyAccess::createPropertyAccessor();
```
Comment (dunglas) [inline src/Api/CachedIdentifiersExtractor.php]: `$propertyAccessor ?? PropertyAccess::createPropertyAccessor();`?

File: src/Api/IdentifiersExtractorInterface.php [hunk 6]
```diff
@@ -0,0 +1,33 @@
+<?php
+
+declare(strict_types=1);
```
Comment (dunglas) [inline src/Api/IdentifiersExtractorInterface.php]: Missing blank line

File: src/Api/IdentifiersExtractorInterface.php [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace ApiPlatform\Core\Api;
+
+/**
+ * Extracts identifiers for a given Resource according to the retrieved Metadata.
+
```
Comment (dunglas) [inline src/Api/IdentifiersExtractorInterface.php]: Line to remove

File: src/Api/IdentifiersExtractorInterface.php [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+ *
+ * @author Antoine Bluchet <soyuka@gmail.com>
+ */
+interface IdentifiersExtractorInterface
+{
+    /**
+     * Find identifiers from an Item (Object).
```
Comment (dunglas) [inline src/Api/IdentifiersExtractorInterface.php]: Finds

File: src/Exception/CacheException.php [hunk 9]
```diff
@@ -0,0 +1,19 @@
+<?php
+
+declare(strict_types=1);
```
Comment (dunglas) [inline src/Exception/CacheException.php]: Blank line

File: tests/Api/CachedIdentifiersExtractorTest.php [hunk 10]
```diff
@@ -0,0 +1,135 @@
+<?php
+
+declare(strict_types=1);
```
Comment (dunglas) [inline tests/Api/CachedIdentifiersExtractorTest.php]: Blank line

File: src/Api/CachedIdentifiersExtractor.php [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+                        continue;
+                    }
+
+                    $relatedItem = $identifiers[$propertyName];
+                    $relatedCacheKey = self::CACHE_KEY_PREFIX.md5($this->getObjectClass($relatedItem));
+
+                    $relatedCacheItem = $this->cacheItemPool->getItem($relatedCacheKey);
+
+                    if (!$relatedCacheItem->isHit()) {
+                        throw new CacheException('No relation cache item founded, we need more cache to continue.');
```
Comment (dunglas) [inline src/Api/CachedIdentifiersExtractor.php]: Can't we avoid to use exceptions? Exception handling is slow.

File: tests/Api/IdentifiersExtractorTest.php [hunk 12]
```diff
@@ -0,0 +1,111 @@
+<?php
+
+declare(strict_types=1);
```
Comment (dunglas) [inline tests/Api/IdentifiersExtractorTest.php]: blank line

**Comment by soyuka** [comment]
@dunglas this should be good now!

File: src/Api/IdentifiersExtractor.php [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+
+    private $propertyNameCollectionFactory;
+    private $propertyMetadataFactory;
+    private $propertyAccessor;
+
+    public function __construct(PropertyNameCollectionFactoryInterface $propertyNameCollectionFactory, PropertyMetadataFactoryInterface $propertyMetadataFactory, PropertyAccessorInterface $propertyAccessor = null)
+    {
+        $this->propertyNameCollectionFactory = $propertyNameCollectionFactory;
+        $this->propertyMetadataFactory = $propertyMetadataFactory;
+        $this->propertyAccessor = $propertyAccessor ?: PropertyAccess::createPropertyAccessor();
```
Comment (dunglas) [inline src/Api/IdentifiersExtractor.php]: `??`

File: src/Bridge/Symfony/Routing/IriConverter.php [hunk 14]
```diff
@@ -42,15 +43,23 @@
...
 
-    public function __construct(PropertyNameCollectionFactoryInterface $propertyNameCollectionFactory, PropertyMetadataFactoryInterface $propertyMetadataFactory, ItemDataProviderInterface $itemDataProvider, RouteNameResolverInterface $routeNameResolver, RouterInterface $router, PropertyAccessorInterface $propertyAccessor = null)
+    public function __construct(PropertyNameCollectionFactoryInterface $propertyNameCollectionFactory, PropertyMetadataFactoryInterface $propertyMetadataFactory, ItemDataProviderInterface $itemDataProvider, RouteNameResolverInterface $routeNameResolver, RouterInterface $router, PropertyAccessorInterface $propertyAccessor = null, IdentifiersExtractorInterface $identifiersExtractor = null)
     {
         $this->propertyNameCollectionFactory = $propertyNameCollectionFactory;
         $this->propertyMetadataFactory = $propertyMetadataFactory;
         $this->itemDataProvider = $itemDataProvider;
         $this->routeNameResolver = $routeNameResolver;
         $this->router = $router;
         $this->propertyAccessor = $propertyAccessor ?: PropertyAccess::createPropertyAccessor();
```
Comment (dunglas) [inline src/Bridge/Symfony/Routing/IriConverter.php]: Can you change this one too? `??`

**Pull request merged**